modules/core/src/test/java/org/apache/ignite/internal/binary/BinaryArraySelfTest.java

```diff
@@ -0,0 +1,348 @@
+        c.put(2, new TestClass(new TestClass1[] {}));
+        c.put(3, new TestClass(new TestClass1[] {new TestClass1() {}}));
+
+        TestClass val1 = (TestClass)c.get(1);
+        TestClass val2 = (TestClass)c.get(2);
+        TestClass val3 = (TestClass)c.get(3);
+
+        assertEquals(2, val1.getArr().length);
+        assertEquals(0, val2.getArr().length);
+        assertEquals(1, val3.getArr().length);
+
+        if (useTypedArrays) {
+            assertTrue(c.replace(1, val1, val2));
+            assertTrue(c.replace(2, val2, val3));
+            assertTrue(c.replace(3, val3, val1));
+        }
+
+        assertTrue(c.remove(1));
+        assertTrue(c.remove(2));
+        assertTrue(c.remove(3));
+    }
+
+    /** */
+    private void doTestBinaryModeArray(IgniteCache<Object, Object> c) {
+        c.put(1, new TestClass1[] {new TestClass1(), new TestClass1()});
+        Object obj = c.withKeepBinary().get(1);
+
+        assertEquals(useTypedArrays ? BinaryArray.class : Object[].class, obj.getClass());
+
+        if (useTypedArrays)
+            assertEquals(TestClass1[].class, ((BinaryObject)obj).deserialize().getClass());
+
+        assertTrue(c.remove(1));
+    }
+
+    /** */
+    private void doTestBoxedPrimitivesArrays(IgniteCache<Object, Object> c) {
+        Object[] data = new Object[] {
+            new Byte[] {1, 2, 3},
+            new Short[] {1, 2, 3},
+            new Integer[] {1, 2, 3},
+            new Long[] {1L, 2L, 3L},
+            new Float[] {1f, 2f, 3f},
+            new Double[] {1d, 2d, 3d},
+            new Character[] {'a', 'b', 'c'},
+            new Boolean[] {true, false},
+        };
+
+        for (Object item : data) {
+            c.put(1, item);
+
+            Object item0 = c.get(1);
+
+            if (useTypedArrays)
+                assertTrue(c.replace(1, item, item));
+
+            assertTrue(c.remove(1));
+
+            if (useTypedArrays)
+                assertEquals(item.getClass(), item0.getClass());
+
+            assertTrue(Arrays.equals((Object[])item, (Object[])item0));
+
+            c.put(item, 1);
+
+            assertEquals(1, c.get(item));
+
+            if (useTypedArrays)
+                assertTrue(c.replace(item, 1, 2));
+
+            assertTrue(c.remove(item));
+        }
+    }
+
+    /** */
+    @Test
+    public void testArrayOfBinariesSerDe() {
+        BinaryObject[] arr = new BinaryObject[] {
+            server.binary().toBinary(new TestClass1()),
+            server.binary().toBinary(new TestClass1())
+        };
+
+        Object obj = server.binary().toBinary(arr);
+
+        Object deser;
+
+        if (useTypedArrays) {
+            assertTrue(obj instanceof BinaryArray);
+
+            deser = ((BinaryArray)obj).deserialize();
+        }
+        else {
+            assertTrue(obj instanceof Object[]);
+
+            deser = PlatformUtils.unwrapBinariesInArray((Object[])obj);
+        }
+
+        assertEquals(Object[].class, deser.getClass());
```

Comment:
if `toBinary` will keep `BinaryObject` array component type then during deserialization on `get` we will got `ArrayStoreException` because component type of the array is `BinaryObject` but element is `TestClass1`.